File: Data Structures/Arrays/Hashing/LeetCode/271Encode/271Encode.py
"""Intuition
After understanding this problem, I wanted to come up with a generic solution that could handle any character. Hence, I tried thinking about ways I can get the original string back without tampering it. I decided to use the lengths and catch strings of given lengths while decoding.

Approach
I start by calculating the total number of strings to be encoded. Then, I calculate the lengths of each strings and store it in a list. I then join the first value and this list with a delimiter. I combine all the strings and prepend the length string to it.
ex: ["a", "bb", "ccc"]
length: 3
lengths: [1, 2, 3]
encoded string: 3,1,2,3,abbccc
Now, while decoding, I separate the string with the delimiter and then reverse engineer the encode function. This approach is similar to a public-secret key approach of encoding.

Complexity
Time complexity:
O(n), n = length of s

Space complexity:
O(n), n = length of s

"""
class Codec:
    def encode(self, strs: List[str]) -> str:
        """Encodes a list of strings to a single string.
        """
        length = len(strs)
        lengths = [str(len(x)) for x in strs]
        return str(length) + "," + ','.join(lengths) + "," + "".join(strs)

    def decode(self, s: str) -> List[str]:
        """Decodes a single string to a list of strings.
        """
        separator = s.split(",")
        length = int(separator[0])
        lengths = []
        for i in range(1, length+1):
            lengths.append(int(separator[i]))
        ignoreIndex = len(','.join(separator[:length+1])) + 1
        jointString = s[ignoreIndex:]
        answer = []
        start, end = 0, 0
        for endIndex in lengths:
            start = end
            end = start+endIndex
            thisString = jointString[start:end]
            answer.append(thisString)
        return answer
# Your Codec object will be instantiated and called as such:
# codec = Codec()
# codec.decode(codec.encode(strs))

File: Data Structures/Arrays/Hashing/LeetCode/271Encode/271Encode.cpp


//Input
/*
We are given a list of strings
*/

//What DS/Algo/Technique
/*
Delimiter-based encoding approach.
*/

//What to do with the data
/*
Encoding
        we have a list of strings
        make them into a single string
Steps:
1. Initialize an empty string 'encoded'
2. Iterate over each string 's' in our list of strings 'strs'
    2.1 We will append the length of string to encoded followed bya. delimiter
    2.2 Append our string (s) to encoded
3. Return encoded

Decoding 
        take the single string
        make the single string into a list of strings

Steps:
1. Initialize an empty vector of strings 'decoded'
2. Initialize an index 'i' to 0
3. While i is less than the length of the encoded string
    3.1 Find the position of the delimiter in the encoded string starting from index 'i'
    3.2 Extract the length of the string from the encoded string based on the delimiter position
    3.3 Move the index i to the position after the delimiter
    3.4 Extract the actual string form the encoded string based on the length
    3.5 Append the extracted string to decoded
    3.6 Update the index i to the position after the extracted string
4. Return the decoded


Creating our delimiter
It will be a private character indicaded as '/'
*/

//Output
/*
We need to fill in the functions for encoding and decoding. 
For encoding we need to return the encoded string. 
For decoding we need to return the lof decoded strings.
*/

#include <vector>
#include <string>

class Codec {
    public:
        // Encodes a list of strings to a single string.
        string encode(vector<string>& strs) {
            //Step 1
            string encoded;
            //Step 2
            for(const string& s: strs){
                //2.1
                encoded += to_string(s.length()) + delimiter;
                //2.2
                encoded += s;
            }
    
            //Step 3
            return encoded;
        }
    
        // Decodes a single string to a list of strings.
        vector<string> decode(string s) {
            //Step 1
            vector<string> decoded;
            //Step 2
            int i = 0;
    
            //Step 3   
            while(i < s.length()){
                //3.1
                size_t delimiterPos = s.find(delimiter, i);
                //3.2
                int length = stoi(s.substr(i, delimiterPos - i));
                //3.3 
                i = delimiterPos + 1;
                //3.4            
                string str = s.substr(i, length);
                //3.5
                decoded.push_back(str);
                //3.6   
                i += length;         
            }   
            //Step 4
            return decoded;
        }
        //Delimiter
        private:
            const char delimiter = '/';
    };
    
    // Your Codec object will be instantiated and called as such:
    // Codec codec;
    // codec.decode(codec.encode(strs));
    
    